### ISSUE TITLE
Add auto-formatting support

### ISSUE THREAD

**Issue by YellowAfterlife** [GameMakerDiscord/GMEdit#1]
### Summary

Several people voiced an opinion that it would have been nice to have an option to auto-format code, such as that in Visual Studio (Ctrl+K, Ctrl+D) or other commonly used code editors.

### Considerations

Ace comes with a [beautify](https://github.com/ajaxorg/ace/blob/master/lib/ace/ext/beautify.js) plugin for this purpose, but [PHP rules](https://github.com/ajaxorg/ace/blob/master/lib/ace/ext/beautify/php_rules.js) do not work well for GML at all, and would likely have to be rewritten more or less completely to account for GML-specific keywords and structures.

In a general case a JS-specific beautifier could be used (which generally work based on parentheses rather than tokens), but these will generally butcher GML preprocessor-like structures (`#define`, `#macro`, etc.) so would require additional logic.

There seems to be an existing [GML-specific beatufier](https://github.com/nstbayless/gml-beautify) but it is written in C++. It hard to estimate whether compiling an Emscripten module for a auto-formatter would be an overkill.

Ace extension route seems to be by far the most logical choice, as relying on Ace token data would be more forward-compatible than writing a custom parser just for this purpose.

### Implementation

This can be implemented either in Haxe or JavaScript (or anything else that compiles to JS) - ultimately the job of the extension/module is to take code as a string and return updated code as a string.

**Comment by christopherwk210** [comment]
One way to solve this issue, and open the door to more possibilities, would be to write a JS compatible GML parser. Might be overkill for this specific task, but it could benefit the growth of the tool overall, and other tools for that matter.

**Comment by YellowAfterlife** [comment]
Parsing is already handled between editor component and GML highlighter. For example, you can do (Ctrl+Shift+I to open devtools)
```js
var TokenIterator = ace.require("ace/token_iterator").TokenIterator;
var iter = new TokenIterator(aceEditor.session, 0, 0);
var tk = iter.getCurrentToken();
while (tk != null) {
	console.log(JSON.stringify(tk));
	tk = iter.stepForward();
}
```
for a document containing
```gml
global.some = 1;
```
and get something like
```js
{"type":"keyword","value":"global","index":0,"start":0}
{"type":"punctuation.operator","value":"."}
{"type":"globalfield","value":"some"}
{"type":"text","value":" "}
{"type":"set.operator","value":"="}
{"type":"text","value":" "}
{"type":"constant.numeric","value":"1"}
{"type":"punctuation.operator","value":";"}
```
(which has all tokens minus the newline characters)

**Comment by christopherwk210** [comment]
Ah, then the only step would be converting the tokens into an AST at that point. Although, for beautifying, it could be done with just the token list, which would probably be the easiest way.

Do you have any plans on implementing a plugin system for the editor? If so, by exposing the token list to a JS plugin, a beautifier could be implemented that way. That way the majority of the base project could still be written in Haxe, instead of splitting up core functionality between languages, if someone like me were to help write this feature (since I know JS and not Haxe).

**Comment by ShinaZin** [comment]
At this time auto-formatting doesn't work correctly with _#import #event_ lines:
![gmedit](https://user-images.githubusercontent.com/18055854/36646729-a5da94d2-1a8c-11e8-9da9-a199ba32a0a0.gif)

**Comment by christopherwk210** [comment]
Fixed in c9f1fd61b255b13bb7ff96443898a860af017a90
